### PR TITLE
STM32 ADC minor cleanup and rework

### PR DESCRIPTION
**Pull request by gautierg-st**
This PR cleans up a bit the code of the STM32 ADC driver:
- Remove some redundant code
- Invert the logic of some #ifdef, to generally reduce length, and make integration of future series easier
- Minor rework of calibration function for F1 and F37x
- Add some missing ifdef for F37x
- Add and fix some comments

### CODE REVIEW

File: drivers/adc/adc_stm32.c [hunk 1]
```diff
@@ -648,27 +648,25 @@ static int adc_stm32_enable(ADC_TypeDef *adc)
...
 	 * still not stabilized, this will wait for a short time (about 1ms)
 	 * to ensure ADC modules are properly enabled.
 	 */
 	uint32_t count_timeout = 0;
 
 	while (LL_ADC_IsActiveFlag_ADRDY(adc) == 0) {
+		/* For F0, continue to write ADEN=1 until ADRDY=1 */
 		if (LL_ADC_IsEnabled(adc) == 0UL) {
 			LL_ADC_Enable(adc);
 		}
```
Comment (erwango) [inline drivers/adc/adc_stm32.c:670]: Should it be F0 specific then ?

File: drivers/adc/adc_stm32.c [hunk 2]
```diff
@@ -1089,30 +1090,21 @@ static int start_read(const struct device *dev,
...
-	defined(CONFIG_SOC_SERIES_STM32G4X) || \
-	defined(CONFIG_SOC_SERIES_STM32H5X) || \
-	defined(CONFIG_SOC_SERIES_STM32H7X) || \
-	defined(CONFIG_SOC_SERIES_STM32U5X) || \
-	defined(CONFIG_SOC_SERIES_STM32WLX)
-	LL_ADC_EnableIT_EOC(adc);
+#if defined(CONFIG_SOC_SERIES_STM32F2X) || \
+	defined(CONFIG_SOC_SERIES_STM32F4X) || \
+	defined(CONFIG_SOC_SERIES_STM32F7X) || \
+	defined(CONFIG_SOC_SERIES_STM32L1X)
```
Comment (erwango) [inline drivers/adc/adc_stm32.c:1101]: Looks like we can start thinking about introducing IP dedicated compatibles with for instance st,stm32f4-adc which could be used by F2/F4/F7/L1 series. Am I right ?

File: drivers/adc/adc_stm32.c (hunk 1 above)
Comment (gautierg-st) [reply drivers/adc/adc_stm32.c:671]: Possibly. I found this note only in F0 reference manual. It is likely not necessary on other series. I'll try making it F0 specific and launching some tests

File: drivers/adc/adc_stm32.c (hunk 2 above)
Comment (gautierg-st) [reply drivers/adc/adc_stm32.c:1101]: Yes, every time these 4 are defined together. This was also the case for the resolution. Sampling time values are different for L1, but that is treated in PR #57342

File: drivers/adc/adc_stm32.c (hunk 1 above)
Comment (gautierg-st) [reply drivers/adc/adc_stm32.c:671]: L1 is also different from F2/F4/F7 regarding the ADC prescaler. But that is a whole separate topic on its own.

**Comment by erwango** [comment]
@xyzzy42, @heinwessels  you might be interested

File: drivers/adc/adc_stm32.c [hunk 3]
```diff
@@ -232,6 +240,12 @@ static const uint32_t table_samp_time[] = {
 /* External channels (maximum). */
 #define STM32_CHANNEL_COUNT		20
 
+/*
+ * Number of ADC clock cycles to wait before starting calibration and after
+ * enabling ADC for STM32F1 and STM32F37x
+ */
+#define STM32F1_ADC_DELAY_CALIB_ENABLE_ADC_CYCLES	2
```
Comment (FRASTM) [inline drivers/adc/adc_stm32.c:247]: this is: 
#define STM32F1_ADC_DELAY_CALIB_ENABLE_ADC_CYCLES  LL_ADC_DELAY_ENABLE_CALIB_ADC_CYCLES

File: drivers/adc/adc_stm32.c [hunk 4]
```diff
@@ -1526,15 +1480,16 @@ static int adc_stm32_init(const struct device *dev)
...
-#if defined(CONFIG_SOC_SERIES_STM32F1X) || \
-	defined(STM32F3X_ADC_V2_5)
-	/*
-	 * Calibration of F1 and F3 (ADC1_V2_5) must starts after two cycles
-	 * after ADON is set.
-	 */
-	LL_ADC_StartCalibration(adc);
+#if defined(CONFIG_SOC_SERIES_STM32F1X) || defined(STM32F3X_ADC_V2_5)
+	wait_cycles = SystemCoreClock / adc_rate *
+		      STM32F1_ADC_DELAY_CALIB_ENABLE_ADC_CYCLES;
```
Comment (FRASTM) [inline drivers/adc/adc_stm32.c:1485]: could adc_rate  be null ?

**Review by FRASTM — COMMENTED** [review]
LGTM

File: drivers/adc/adc_stm32.c (hunk 3 above)
Comment (gautierg-st) [reply drivers/adc/adc_stm32.c:247]: Correct, I didn't see that. There is yet another macro for STM32F37x. I'll update this part

File: drivers/adc/adc_stm32.c (hunk 4 above)
Comment (gautierg-st) [reply drivers/adc/adc_stm32.c]: It clearly shouldn't be, but in doubt, I added a check. Also moved this waiting in a specific function.

**Review by erwango — DISMISSED** [review]

File: dts/arm/st/f1/stm32f1.dtsi [hunk 5]
```diff
@@ -333,7 +333,7 @@
 		};
 
 		adc1: adc@40012400 {
-			compatible = "st,stm32-adc";
+			compatible = "st,stm32-adc", "st,stm32f1-adc";
```
Comment (trentp-igor) [inline dts/arm/st/f1/stm32f1.dtsi:336]: The normal convention is for the most specific compatible to be listed first and then to get more generic.  So stm32f1-adc would be first, then stm32-adc.

Comment (gautierg-st) [reply dts/arm/st/f1/stm32f1.dtsi]: Thanks, I wasn't aware of this convention. Fixed now.

**Review by erwango — DISMISSED** [review]

**Review by erwango — DISMISSED** [review]

**Review by erwango — APPROVED** [review]

**Review by FRASTM — APPROVED** [review]

**Pull request merged**